--- src/dird/ua_cmds.c.orig 2005-04-06 15:19:36.636609424 -0400
+++ src/dird/ua_cmds.c 2005-04-06 15:20:13.176649111 -0400
@@ -98,7 +98,7 @@
  { N_("add"),        add_cmd,         _("add media to a pool")},
  { N_("autodisplay"), autodisplay_cmd, _("autodisplay [on|off] -- console messages")},
  { N_("automount"),   automount_cmd,  _("automount [on|off] -- after label")},
- { N_("cancel"),     cancel_cmd,    _("cancel <job=nnn> -- cancel a job")},
+ { N_("cancel"),     cancel_cmd,    _("cancel <jobid=nnn> <job=name>-- cancel a job")},
  { N_("create"),     create_cmd,    _("create DB Pool from resource")},  
  { N_("delete"),     delete_cmd,    _("delete [pool=<pool-name> | media volume=<volume-name>]")},    
  { N_("estimate"),   estimate_cmd,  _("performs FileSet estimate, listing gives full listing")},
