--- src/cats/make_postgresql_tables.in.orig	2005-02-27 00:24:56.000000000 -0500
+++ src/cats/make_postgresql_tables.in	2004-12-05 15:16:16.000000000 -0500
@@ -212,14 +212,9 @@
 
 create index file_jobid_idx on file (jobid);
 create index file_fp_idx on file (filenameid, pathid);
-
---
--- Possibly add one or more of the following indexes
---  if your Verifies are too slow.
---
--- create index file_pathid_idx on file(pathid);
--- create index file_filenameid_idx on file(filenameid);
--- create index file_jpfid_idx on file (jobid, pathid, filenameid);
+create index file_pathid_idx on file(pathid);
+create index file_filenameid_idx on file(filenameid);
+create index file_jpfid_idx on file (jobid, pathid, filenameid);
 
 create table jobmedia
 (
