--- src/dird/bacula-dir.conf.in.old	2008-11-26 14:17:16.000000000 +0100
+++ src/dird/bacula-dir.conf.in	2008-11-26 14:18:12.000000000 +0100
@@ -62,11 +62,17 @@
   FileSet="Catalog"
   Schedule = "WeeklyCycleAfterBackup"
   # This creates an ASCII copy of the catalog
+  #
   # WARNING!!! Passing the password via the command line is insecure.
   # see comments in make_catalog_backup for details.
   # Arguments to make_catalog_backup are:
   #  make_catalog_backup <database-name> <user-name> <password> <host>
-  RunBeforeJob = "@scriptdir@/make_catalog_backup @db_name@ @db_user@"
+  #
+  # Ubuntu uses make_catalog_backup_awk script for
+  # security reasons
+  # Replace <CatalogName> with the real Catalog name
+  #
+  RunBeforeJob = "/usr/bin/awk -f @scriptdir@/make_catalog_backup_awk -v cat1=<CatalogName> @sysconfdir@/bacula-dir.conf"
   # This deletes the copy of the catalog
   RunAfterJob  = "@scriptdir@/delete_catalog_backup"
   Write Bootstrap = "@working_dir@/BackupCatalog.bsr"
