--- scripts/mtx-changer.in.old	2007-08-16 15:01:19.000000000 +0200
+++ scripts/mtx-changer.in	2008-04-04 23:13:55.000000000 +0200
@@ -2,7 +2,7 @@
 #
 # Bacula interface to mtx autoloader
 #
-#  $Id: mtx-changer.in 5360 2007-08-16 13:01:19Z pbuschman $
+#  $Id: mtx-changer.in 6712 2008-04-01 10:27:46Z kerns $
 #
 #  If you set in your Device resource
 #
@@ -52,6 +52,10 @@
 # SunOS     No Additional Sense
 # FreeBSD   Current Driver State: at rest.
 # Linux     ONLINE
+#  Note Debian has a different mt than the standard Linux version. 
+#    When no tape is in the drive it waits 2 minutes.  
+#    When a tape is in the drive, it prints user unfriendly output.
+#
 
 OS=`uname`
 case ${OS} in
@@ -61,8 +65,12 @@
   FreeBSD)
     ready="Current Driver State: at rest."
     ;;
-  *)
-    ready="ONLINE"
+  Linux)
+    if test -e /etc/debian_version ; then
+       ready="drive status"
+    else
+       ready="ONLINE"
+    fi
   ;;
 esac
 
@@ -108,7 +116,7 @@
 wait_for_drive() {
   i=0 
   while [ $i -le 300 ]; do  # Wait max 300 seconds
-    if mt -f $1 status | grep "${ready}" >/dev/null 2>&1; then
+    if mt -f $1 status 2>&1 | grep "${ready}" >/dev/null 2>&1; then
       break
     fi
     debug "Device $1 - not ready, retrying..."
