diff -Naur bacula-2.4.2.orig/src/cats/grant_postgresql_privileges.in bacula-2.4.2/src/cats/grant_postgresql_privileges.in
--- bacula-2.4.2.orig/src/cats/grant_postgresql_privileges.in	2008-08-04 10:20:20.000000000 -0400
+++ bacula-2.4.2/src/cats/grant_postgresql_privileges.in	2008-08-25 11:11:39.000000000 -0400
@@ -2,9 +2,13 @@
 #
 # shell script to grant privileges to the bacula database
 #
-USER=@db_user@
+if [ ! -n "$USER" ]; then
+	USER=@db_user@
+fi
 bindir=@SQL_BINDIR@
-db_name=@db_name@
+if [ ! -n "$db_name" ]; then
+	db_name=@db_name@
+fi
 
 if $bindir/psql -f - -d ${db_name} $* <<END-OF-DATA
 
