Index: src/stored/mount.c
===================================================================
--- src/stored/mount.c	(revision 6978)
+++ src/stored/mount.c	(working copy)
@@ -290,7 +290,9 @@
       if (!is_eod_valid()) {
          goto mount_next_vol;
       }
-
+      if (!dev->VolCatInfo.VolFirstWritten) {
+	 dev->VolCatInfo.VolFirstWritten = time(NULL);
+      }
       dev->VolCatInfo.VolCatMounts++;      /* Update mounts */
       Dmsg1(150, "update volinfo mounts=%d\n", dev->VolCatInfo.VolCatMounts);
       if (!dir_update_volume_info(dcr, false, false)) {
