
This patch improve a bit performance during attribute despooling


Index: src/dird/getmsg.c
===================================================================
--- src/dird/getmsg.c	(révision 8258)
+++ src/dird/getmsg.c	(copie de travail)
@@ -198,6 +198,11 @@
          continue;
       }
 
+      if (bs->msg[0] == 'U') {        /* SD sending attributes */
+         Dmsg2(900, "Catalog upd jcr 0x%x: %s", jcr, bs->msg);
+         catalog_update(jcr, bs);
+         continue;
+      }
       /*
        * Here we are expecting a message of the following format:
        *   Jmsg Job=nnn type=nnn level=nnn Message-string
@@ -231,11 +236,6 @@
          catalog_request(jcr, bs);
          continue;
       }
-      if (bs->msg[0] == 'U') {        /* SD sending attributes */
-         Dmsg2(900, "Catalog upd jcr 0x%x: %s", jcr, bs->msg);
-         catalog_update(jcr, bs);
-         continue;
-      }
       if (bs->msg[0] == 'M') {        /* Mount request */
          Dmsg1(900, "Mount req: %s", bs->msg);
          mount_request(jcr, bs, msg);
