Index: scripts/cleanup-tape.in
===================================================================
--- scripts/cleanup-tape.in	(révision 7994)
+++ scripts/cleanup-tape.in	(copie de travail)
@@ -12,16 +12,16 @@
 init_slot @tape_drive@ $SLOT1
 
 if test x@autochanger@ != x/dev/null; then
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive@ $DRIVE1`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive@ $DRIVE1`
    case $slot in
    0)
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
       ;;
    $SLOT1)
       ;;
    *)
-      bin/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
       ;;
   esac
 fi
Index: scripts/prepare-two-tapes.in
===================================================================
--- scripts/prepare-two-tapes.in	(révision 7994)
+++ scripts/prepare-two-tapes.in	(copie de travail)
@@ -18,10 +18,10 @@
 # optimize so we don't unnecessarily unload and load slots
 #
 if test x@autochanger@ != x/dev/null; then
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive@ $DRIVE1`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive@ $DRIVE1`
    case $slot in
    0)
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
       slot=$SLOT1
       ;;
    $SLOT1)
@@ -31,22 +31,22 @@
       slot=$SLOT2
       ;;
    *)
-      bin/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
       slot=$SLOT1
       ;;
   esac
   init_drive @tape_drive@
-  bin/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
+  $scripts/@changer_script@ @autochanger@ unload $slot @tape_drive@ $DRIVE1
 
 
 # init the other slot
   case $slot in
    $SLOT1)
-      bin/@changer_script@ @autochanger@ load $SLOT2 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT2 @tape_drive@ $DRIVE1
       ;;
    $SLOT2)
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ $DRIVE1
       ;;
    *)
       echo "Something went wrong. Expected $SLOT1 or $SLOT2, got $slot"
Index: scripts/check_for_zombie_jobs
===================================================================
--- scripts/check_for_zombie_jobs	(révision 7994)
+++ scripts/check_for_zombie_jobs	(copie de travail)
@@ -14,7 +14,7 @@
     client=${HOST}-fd
 fi
 
-bin/bconsole -c bin/bconsole.conf <<END_OF_DATA 2>&1 >/dev/null
+$bin/bconsole -c bin/bconsole.conf <<END_OF_DATA 2>&1 >/dev/null
 @output tmp/dir.out
 status dir
 @output tmp/fd.out
Index: scripts/prepare-other-loc
===================================================================
--- scripts/prepare-other-loc	(révision 8011)
+++ scripts/prepare-other-loc	(copie de travail)
@@ -5,8 +5,8 @@
 SCR=$cwd/bin/bacula-ctl
 sed "s:BACDIRBIN=$cwd/bin:BACDIRBIN=$bin:" $SCR-dir > tmp/1
 mv tmp/1 $SCR-dir
-se   d -i "s:BACSDBIN=$cwd/bin:BACSDBIN=$bin:" $SCR-sd > tmp/1
-mv tm  p/1 $SCR-sd
+sed -i "s:BACSDBIN=$cwd/bin:BACSDBIN=$bin:" $SCR-sd > tmp/1
+mv tmp/1 $SCR-sd
 sed -i "s:BACFDBIN=$cwd/bin:BACFDBIN=$bin:" $SCR-fd > tmp/1
 mv tmp/1 $SCR-fd
 chmod -x $cwd/bin/bacula-{fd,sd,dir} $cwd/bin/bconsole    
Index: scripts/setup
===================================================================
--- scripts/setup	(révision 7994)
+++ scripts/setup	(copie de travail)
@@ -20,6 +20,7 @@
    exit 1
 fi
 
+# we build and setup binaries to the $cwd/bin directory not $bin
 rm -rf build bin
 # Copy new source
 echo "Copying source from ${BACULA_SOURCE}"
Index: scripts/cleanup-2drive.in
===================================================================
--- scripts/cleanup-2drive.in	(révision 7994)
+++ scripts/cleanup-2drive.in	(copie de travail)
@@ -6,40 +6,40 @@
 scripts/cleanup
 
 if test x@autochanger@ != x/dev/null; then
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive1@ 1`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive1@ 1`
    echo "Slot $slot in drive 1"
    if test x$slot != x0 ; then
       echo "unload slot $slot in drive 1"
-      bin/@changer_script@ @autochanger@ unload $slot @tape_drive1@ 1
+      $scripts/@changer_script@ @autochanger@ unload $slot @tape_drive1@ 1
    fi
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive@ 0`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive@ 0`
    echo "Slot $slot in drive 0"
    if test x$slot != x0 ; then
       echo "unload slot $slot in drive 0"
-      bin/@changer_script@ @autochanger@ unload $slot @tape_drive@ 0
+      $scripts/@changer_script@ @autochanger@ unload $slot @tape_drive@ 0
    fi
 
 
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive1@ 1`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive1@ 1`
    if test x$slot != x$SLOT2; then
       echo "load slot $SLOT2 in drive 1"
-      bin/@changer_script@ @autochanger@ load $SLOT2 @tape_drive1@ 1
+      $scripts/@changer_script@ @autochanger@ load $SLOT2 @tape_drive1@ 1
    fi
    echo "WEOF drive 1"
    init_drive @tape_drive1@
 
    echo "unload slot $SLOT2 in drive 1"
-   bin/@changer_script@ @autochanger@ unload $SLOT2 @tape_drive1@ 1
+   $scripts/@changer_script@ @autochanger@ unload $SLOT2 @tape_drive1@ 1
 
-   slot=`bin/@changer_script@ @autochanger@ loaded 0 @tape_drive@ 0`
+   slot=`$scripts/@changer_script@ @autochanger@ loaded 0 @tape_drive@ 0`
    echo "Slot $slot in drive 0"
    if test x$slot != x$SLOT1; then
       echo "load slot $SLOT1 in drive 0"
-      bin/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ 0
+      $scripts/@changer_script@ @autochanger@ load $SLOT1 @tape_drive@ 0
    fi
    echo "WEOF drive 0"
    init_drive @tape_drive@
 
    echo "unload slot $SLOT1 from drive 0"
-   bin/@changer_script@ @autochanger@ unload $SLOT1 @tape_drive@ 0
+   $scripts/@changer_script@ @autochanger@ unload $SLOT1 @tape_drive@ 0
 fi
Index: scripts/functions
===================================================================
--- scripts/functions	(révision 7994)
+++ scripts/functions	(copie de travail)
@@ -10,11 +10,11 @@
    echo "s%  mail =%# mail = %g" >${outf}
    echo "s%  operator =%# operator =%g" >>${outf}
    cp bin/bacula-dir.conf tmp/1
-   sed -f ${outf} tmp/1 >bin/bacula-dir.conf
+   sed -f ${outf} tmp/1 > bin/bacula-dir.conf
    echo " "
    echo " "
    echo " === Starting $TestName at `date +%R:%S` ==="
-   echo " === Starting $TestName at `date +%R:%S` ===" >>working/log
+   echo " === Starting $TestName at `date +%R:%S` ===" >> working/log
    echo " "
    export zstat
 }
@@ -147,11 +147,11 @@
      bin/bacula-ctl-sd start
      bin/bacula-ctl-fd start $1
      bin/bacula-ctl-dir start
-     cat tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
+     cat tmp/bconcmds | $bin/bconsole -c bin/bconsole.conf
      return $?
    else
-     bin/bacula start 2>&1 >/dev/null
-     cat tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
+     bin/bacula start >/dev/null 2>&1
+     cat tmp/bconcmds | $bin/bconsole -c bin/bconsole.conf >/dev/null 2>&1
      return $?
    fi
 }
@@ -159,27 +159,27 @@
 run_bconsole()
 {
    if test "$debug" -eq 1 ; then
-     cat tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
+     cat tmp/bconcmds | $bin/bconsole -c bin/bconsole.conf
    else
-     cat tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
+     cat tmp/bconcmds | $bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
    fi
 }
 
 run_btape()
 {
    if test "$debug" -eq 1 ; then
-     cat tmp/bconcmds | bin/btape -c bin/bacula-sd.conf DDS-4 | tee tmp/log1.out
+     cat tmp/bconcmds | $bin/btape -c bin/bacula-sd.conf DDS-4 | tee tmp/log1.out
    else
-     cat tmp/bconcmds | bin/btape -c bin/bacula-sd.conf DDS-4 2>&1 >tmp/log1.out
+     cat tmp/bconcmds | $bin/btape -c bin/bacula-sd.conf DDS-4 >tmp/log1.out 2>&1
    fi
 }
 
 run_bscan()
 {
    if test "$debug" -eq 1 ; then
-      bin/bscan $* | tee tmp/log.out
+      $bin/bscan $* | tee tmp/log.out
    else
-      bin/bscan $* 2>&1 >/dev/null
+      $bin/bscan $* 2>&1 >/dev/null
    fi
 }
 
@@ -356,10 +356,10 @@
 load_slot1() 
 {
 # Get a tape from slot1
-slot=`bin/$MTX ${AUTOCHANGER} loaded 0 ${TAPE_DRIVE} $DRIVE1`
+slot=`$scripts/$MTX ${AUTOCHANGER} loaded 0 ${TAPE_DRIVE} $DRIVE1`
 case $slot in
  0)
-    bin/$MTX ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/$MTX ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
     slot=$SLOT1
     ;;
  $SLOT1)
@@ -367,8 +367,8 @@
     ;;
  *)
     rewind_drive ${TAPE_DRIVE}
-    bin/$MTX ${AUTOCHANGER} unload $slot  ${TAPE_DRIVE} $DRIVE1
-    bin/$MTX ${AUTOCHANGER} load   $SLOT1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/$MTX ${AUTOCHANGER} unload $slot  ${TAPE_DRIVE} $DRIVE1
+    $scripts/$MTX ${AUTOCHANGER} load   $SLOT1 ${TAPE_DRIVE} $DRIVE1
     slot=$SLOT1
     ;;
 esac
@@ -383,17 +383,17 @@
 rewind_drive ${TAPE_DRIVE}
 case $1 in
  0)
-    bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
     slot=1
     ;;
  $SLOT1)
-    bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $1 ${TAPE_DRIVE} $DRIVE1
-    bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT2 ${TAPE_DRIVE} $DRIVE1
+    $scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT2 ${TAPE_DRIVE} $DRIVE1
     slot=2
     ;;
  $SLOT2)
-    bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $1 ${TAPE_DRIVE} $DRIVE1
-    bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $1 ${TAPE_DRIVE} $DRIVE1
+    $scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} load $SLOT1 ${TAPE_DRIVE} $DRIVE1
     slot=1
     ;;
  *)
@@ -403,7 +403,6 @@
 esac
 }
 
-
 # Save current directory
 cwd=`pwd`
 if test "x${REGRESS_DEBUG}" = "x1"; then
@@ -418,16 +417,23 @@
 db_name=${db_name:-"regress"}
 db_user=${db_user:-"regress"}
 db_password=${db_password:-""}
+working=${working:-"$cwd/working"}
+bin=${bin:-"$cwd/bin"}
+scripts=${scripts:-"$bin"}
 
+export bin
+export working
+export scripts
+
 CLIENT=${HOST}-fd
 
 if [ x$USE_VTAPE = xyes ]; then
-   mkdir -p $cwd/working/ach
+   mkdir -p $working/ach
    SLOT1=1
    SLOT2=2
-   TAPE_DRIVE=$cwd/working/ach/drive1
-   TAPE_DRIVE1=$cwd/working/ach/drive1
-   AUTOCHANGER=$cwd/working/ach/config
+   TAPE_DRIVE=$working/ach/drive1
+   TAPE_DRIVE1=$working/ach/drive1
+   AUTOCHANGER=$working/ach/config
    AUTOCHANGER_SCRIPT=disk-changer
    DRIVE1=0
    DRIVE2=1
@@ -438,6 +444,6 @@
 fi
 
 AUTOCHANGER_SCRIPT=${AUTOCHANGER_SCRIPT:-mtx-changer}
-LD_LIBRARY_PATH=$cwd/bin
+LD_LIBRARY_PATH=$bin
 export LD_LIBRARY_PATH
 
Index: tests/btape-fill-full-tape
===================================================================
--- tests/btape-fill-full-tape	(révision 7994)
+++ tests/btape-fill-full-tape	(copie de travail)
@@ -14,7 +14,7 @@
 change_jobname $JobName
 start_test
 
-bin/btape -c bin/bacula-sd.conf DDS-4 <<END_OF_DATA 2>&1 >${cwd}/tmp/log1.out
+$bin/btape -c bin/bacula-sd.conf DDS-4 <<END_OF_DATA >${cwd}/tmp/log1.out 2>&1
 fill
 s
 
@@ -22,7 +22,7 @@
 END_OF_DATA
 
 
-grep "^The last block on the tape matches\. Test succeeded\." ${cwd}/tmp/log1.out 2>&1 >/dev/null
+grep "^The last block on the tape matches\. Test succeeded\." ${cwd}/tmp/log1.out >/dev/null 2>&1
 if [ $? != 0 ] ; then
    echo " "
    echo " "
Index: tests/disk-changer-test
===================================================================
--- tests/disk-changer-test	(révision 7994)
+++ tests/disk-changer-test	(copie de travail)
@@ -3,43 +3,43 @@
 mkdir tmp/disk-changer
 touch tmp/disk-changer/conf 
 echo "Unload drive 0"
-bin/disk-changer tmp/disk-changer/conf unload 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf unload 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Test what is in drive 0"
-bin/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Load Slot 1 into drive 0"
-bin/disk-changer tmp/disk-changer/conf load 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf load 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Test what is in drive 0"
-bin/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 
 echo "Unload drive 1"
-bin/disk-changer tmp/disk-changer/conf unload 2 tmp/disk-changer/drive1 1
+$scripts/disk-changer tmp/disk-changer/conf unload 2 tmp/disk-changer/drive1 1
 echo "rtn=$?"   
 echo "Test what is in drive 1"
-bin/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive1 1
+$scripts/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive1 1
 echo "rtn=$?"   
 echo "Load Slot 2 drive 1"
-bin/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive1 1
+$scripts/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive1 1
 echo "rtn=$?"   
 echo "Test what is in drive 1"
-bin/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive1 1
+$scripts/disk-changer tmp/disk-changer/conf loaded 1 tmp/disk-changer/drive1 1
 echo "rtn=$?"   
 
 echo "Load Slot 1 into drive 0 -- error"
-bin/disk-changer tmp/disk-changer/conf load 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf load 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Load Slot 2 into drive 1 -- error"
-bin/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive1 1
+$scripts/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive1 1
 echo "rtn=$?"   
 echo "Load Slot 2 in drive 0 -- error"
-bin/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Unload drive 0"
-bin/disk-changer tmp/disk-changer/conf unload 1 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf unload 1 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
 echo "Load Slot 2 in drive 0 -- should fail"
-bin/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive0 0
+$scripts/disk-changer tmp/disk-changer/conf load 2 tmp/disk-changer/drive0 0
 echo "rtn=$?"   
Index: tests/bug-fatal-test
===================================================================
--- tests/bug-fatal-test	(révision 7994)
+++ tests/bug-fatal-test	(copie de travail)
@@ -34,12 +34,11 @@
 END_OF_DATA
 
 
-bin/bacula start
-cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf > /dev/null
+run_bacula
 stop_bacula
 
-export dstat=0
-export bstat=0
-export rstat=0
+dstat=0
+bstat=0
+rstat=0
 
 end_test
Index: tests/errors-test
===================================================================
--- tests/errors-test	(révision 7994)
+++ tests/errors-test	(copie de travail)
@@ -64,7 +64,7 @@
 
 run_bacula
 touch ${cwd}/tmp/log1.out
-client=$(grep client= bin/bacula-dir.conf)
+client=`grep client= $bin/bacula-dir.conf`
 check_for_zombie_jobs storage=File $client
 stop_bacula
 
Index: tests/restore-by-file-tape
===================================================================
--- tests/restore-by-file-tape	(révision 7994)
+++ tests/restore-by-file-tape	(copie de travail)
@@ -93,7 +93,7 @@
 check_two_logs
 
 scripts/check_for_zombie_jobs storage=DDS-4
-bin/bacula stop 2>&1 >/dev/null
+stop_bacula
 for i in `cat ${cwd}/tmp/restore2-list`; do
    diff $i ${cwd}/tmp/bacula-restores$i
    if [ $? != 0 ] ; then
Index: tests/vtape-test-changer
===================================================================
--- tests/vtape-test-changer	(révision 7994)
+++ tests/vtape-test-changer	(copie de travail)
@@ -30,7 +30,7 @@
 when1=`perl -MPOSIX -e "print strftime('%F %T', localtime(time+30))"`
 when2=`perl -MPOSIX -e "print strftime('%F %T', localtime(time+45))"`
 
-clientname=`awk '/Name = .*-fd/ { if (!ok) { print $3 ; ok=1 } }' bin/bacula-dir.conf`
+clientname=`awk '/Name = .*-fd/ { if (!ok) { print $3 ; ok=1 } }' $bin/bacula-dir.conf`
 
 # Catalog record for cleaning tape "CLN01" successfully created.
 # CLN01      | Cleaning
Index: tests/usr-tape-root
===================================================================
--- tests/usr-tape-root	(révision 7994)
+++ tests/usr-tape-root	(copie de travail)
@@ -15,8 +15,7 @@
 echo " "
 echo " "
 
-bin/bacula start 2>&1 >/dev/null
-bin/bconsole -c bin/bconsole.conf <<END_OF_DATA
+cat <<END_OF_DATA > ${cwd}/tmp/bconcmds
 @output /dev/null
 messages
 @output tmp/log1.out
@@ -35,7 +34,8 @@
 @output
 quit
 END_OF_DATA
-bin/bacula stop 2>&1 >/dev/null
+run_bacula
+stop_bacula
 cd /
 ${cwd}/bin/testls -e ${cwd}/scripts/exclude-usr-test lib >${cwd}/tmp/original
 cd ${cwd}/tmp/bacula-restores
Index: tests/concurrent-jobs-test
===================================================================
--- tests/concurrent-jobs-test	(révision 7994)
+++ tests/concurrent-jobs-test	(copie de travail)
@@ -16,7 +16,7 @@
    dd if=/dev/urandom of=${cwd}/tmp/largefile bs=1024 count=55000
 else
    echo "Creating a 56MB file with bacula-dir data ..."
-   dd if=bin/bacula-dir of=${cwd}/tmp/1 bs=1024 count=1000
+   dd if=$bin/bacula-dir of=${cwd}/tmp/1 bs=1024 count=1000
    cat ${cwd}/tmp/1 ${cwd}/tmp/1 ${cwd}/tmp/1 ${cwd}/tmp/1 ${cwd}/tmp/1 >${cwd}/tmp/2
    rm -f ${cwd}/tmp/1
    cat ${cwd}/tmp/2 ${cwd}/tmp/2 ${cwd}/tmp/2 ${cwd}/tmp/2 ${cwd}/tmp/2 >>${cwd}/tmp/3
Index: tests/accurate-test
===================================================================
--- tests/accurate-test	(révision 7994)
+++ tests/accurate-test	(copie de travail)
@@ -157,7 +157,7 @@
 # Check with bls
 ################################################################
 
-${cwd}/bin/bls -c ${cwd}/bin/bacula-sd.conf -V 'TestVolume001' FileStorage > tmp/bls.out
+$bin/bls -c bin/bacula-sd.conf -V 'TestVolume001' FileStorage > tmp/bls.out
 grep -- '----' tmp/bls.out | grep xxx > /dev/null
 if [ $? != 0 ] ; then
     bstat=2
@@ -283,7 +283,7 @@
 
 bscan_libdbi
 
-${cwd}/bin/bscan -c ${cwd}/bin/bacula-sd.conf $BSCANLIBDBI -n regress -u regress -m -s -b tmp/bscan.bsr FileStorage 2>&1 > ${cwd}/tmp/bscan.log
+$bin/bscan -c bin/bacula-sd.conf $BSCANLIBDBI -n regress -u regress -m -s -b tmp/bscan.bsr FileStorage 2>&1 > ${cwd}/tmp/bscan.log
 
 cat <<END_OF_DATA >${cwd}/tmp/bconcmds
 @$out ${cwd}/tmp/log1.out
Index: tests/incremental-2disk
===================================================================
--- tests/incremental-2disk	(révision 7994)
+++ tests/incremental-2disk	(copie de travail)
@@ -9,14 +9,7 @@
 JobName=Inc2disk
 . scripts/functions
 
-
-stop_bacula
-cd bin
-./drop_bacula_tables >/dev/null 2>&1
-./make_bacula_tables >/dev/null 2>&1
-./grant_bacula_privileges 2>&1 >/dev/null
-cd ..
-
+scripts/cleanup
 scripts/copy-2disk-confs
 scripts/prepare-disk-changer
 
Index: tests/lib-tape-root
===================================================================
--- tests/lib-tape-root	(révision 7994)
+++ tests/lib-tape-root	(copie de travail)
@@ -5,13 +5,7 @@
 #
 cwd=`pwd`
 . scripts/functions
-bin/bacula stop 2>&1 >/dev/null
-cd bin
-./drop_bacula_tables >/dev/null 2>&1
-./make_bacula_tables >/dev/null 2>&1
-./grant_bacula_privileges 2>&1 >/dev/null
-cd ..
-
+scripts/cleanup
 scripts/cleanup-tape
 scripts/copy-tape-confs
 echo "/lib" >${cwd}/tmp/file-list
@@ -20,8 +14,8 @@
 echo " === Starting lib-tape-root test ==="
 echo " "
 echo " "
-bin/bacula start 2>&1 >/dev/null
-bin/bconsole -c bin/bconsole.conf <<END_OF_DATA
+
+cat <<END_OF_DATA > >${cwd}/tmp/bconcmds
 @output /dev/null
 messages
 @output ${cwd}/tmp/log1.out
@@ -40,7 +34,10 @@
 @output
 quit
 END_OF_DATA
-bin/bacula stop 2>&1 >/dev/null
+
+run_bacula
+stop_bacula
+
 cd /
 ${cwd}/bin/testls -e ${cwd}/scripts/exclude-usr-test lib >${cwd}/tmp/original
 cd ${cwd}/tmp/bacula-restores
Index: tests/two-volume-changer
===================================================================
--- tests/two-volume-changer	(révision 7994)
+++ tests/two-volume-changer	(copie de travail)
@@ -54,7 +54,7 @@
 check_for_zombie_jobs storage=DDS-4
 stop_bacula
 
-bin/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $SLOT2 ${TAPE_DRIVE} 0
+$scripts/${AUTOCHANGER_SCRIPT} ${AUTOCHANGER} unload $SLOT2 ${TAPE_DRIVE} 0
 
 cat <<END_OF_DATA >tmp/bconcmds
 @$out /dev/null
Index: tests/bextract-test
===================================================================
--- tests/bextract-test	(révision 7994)
+++ tests/bextract-test	(copie de travail)
@@ -45,9 +45,9 @@
 
 mkdir -p ${cwd}/tmp/bacula-restores
 if test "$debug" -eq 1 ; then
-  bin/bextract -v -b working/restore.bsr -c bin/bacula-sd.conf ${cwd}/tmp ${cwd}/tmp/bacula-restores                
+  $bin/bextract -v -b working/restore.bsr -c bin/bacula-sd.conf ${cwd}/tmp ${cwd}/tmp/bacula-restores                
 else
-  bin/bextract -b working/restore.bsr -c bin/bacula-sd.conf ${cwd}/tmp ${cwd}/tmp/bacula-restores 2>&1 >/dev/null
+  $bin/bextract -b working/restore.bsr -c bin/bacula-sd.conf ${cwd}/tmp ${cwd}/tmp/bacula-restores 2>&1 >/dev/null
 fi
 rstat=$?
 grep "^  Termination: *Backup OK" ${cwd}/tmp/log1.out 2>&1 >/dev/null
Index: tests/four-jobs-tape
===================================================================
--- tests/four-jobs-tape	(révision 7994)
+++ tests/four-jobs-tape	(copie de travail)
@@ -30,13 +30,7 @@
 messages
 quit
 END_OF_DATA
-if test "$debug" -eq 1 ; then
-  bin/bacula start
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
-else
-  bin/bacula start 2>&1 >/dev/null
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
-fi
+run_bacula
 
 scripts/check_for_zombie_jobs storage=DDS-4
 echo "Backup 1 done"
@@ -56,14 +50,9 @@
 messages
 quit
 END_OF_DATA
-if test "$debug" -eq 1 ; then
-  bin/bacula start
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
-else
-  bin/bacula start 2>&1 >/dev/null
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
-fi
 
+run_bacula
+
 scripts/check_for_zombie_jobs storage=DDS-4
 echo "Backup 2 done"
 touch ${cwd}/build/src/dird/*.c
@@ -80,14 +69,9 @@
 messages
 quit
 END_OF_DATA
-if test "$debug" -eq 1 ; then
-  bin/bacula start
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
-else
-  bin/bacula start 2>&1 >/dev/null
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
-fi
 
+run_bacula
+
 scripts/check_for_zombie_jobs storage=DDS-4
 echo "Backup 3 done"
 # make some files for the incremental to pick up
@@ -104,13 +88,7 @@
 messages
 quit
 END_OF_DATA
-if test "$debug" -eq 1 ; then
-  bin/bacula start
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf
-else
-  bin/bacula start 2>&1 >/dev/null
-  cat ${cwd}/tmp/bconcmds | bin/bconsole -c bin/bconsole.conf  2>&1 >/dev/null
-fi
+run_bacula
 
 scripts/check_for_zombie_jobs storage=DDS-4
 echo "Backup 4 done"
Index: tests/bscan-test
===================================================================
--- tests/bscan-test	(révision 7994)
+++ tests/bscan-test	(copie de travail)
@@ -67,9 +67,9 @@
 fi
 
 if test "$debug" -eq 1 ; then
-  bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf ${cwd}/tmp                   
+  $bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf ${cwd}/tmp                   
 else
-  bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf ${cwd}/tmp 2>&1 >tmp/log3.out
+  $bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf ${cwd}/tmp >tmp/log3.out 2>&1
 fi
 
 cat <<END_OF_DATA >tmp/bconcmds
Index: tests/bscan-fast-tape
===================================================================
--- tests/bscan-fast-tape	(révision 7994)
+++ tests/bscan-fast-tape	(copie de travail)
@@ -34,7 +34,7 @@
 END_OF_DATA
 
 echo "Starting Bacula tape writing"
-#bin/btape -c bin/bacula-sd.conf /dev/nst0 <<END_OF_DATA
+#$bin/btape -c bin/bacula-sd.conf /dev/nst0 <<END_OF_DATA
 #rewind
 #label
 #Test001
@@ -66,7 +66,7 @@
 
 bscan_libdbi
 
-strace -o strace.new bin/bscan -d200 -w working $BSCANLIBDBI -u regress -n regress -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4
+strace -o strace.new $bin/bscan -d200 -w working $BSCANLIBDBI -u regress -n regress -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4
 exit
 cat <<END_OF_DATA >tmp/bconcmds
 @$out /dev/null
Index: tests/bscan-tape
===================================================================
--- tests/bscan-tape	(révision 7994)
+++ tests/bscan-tape	(copie de travail)
@@ -106,9 +106,9 @@
 fi
 
 if test "$debug" -eq 1 ; then
-  bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4                
+  $bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4                
 else
-  bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4 2>&1 >/dev/null
+  $bin/bscan -w working $BSCANLIBDBI -u ${db_user} -n ${db_name} $PASSWD -m -s -v -b tmp/bscan.bsr -c bin/bacula-sd.conf DDS-4 >/dev/null 2>&1
 fi
 
 echo $BSCANLIBDBI
