diff -Naur --exclude=.svn --exclude='*~' bacula/src/dird/ua_purge.c bacula.new/src/dird/ua_purge.c
--- bacula/src/dird/ua_purge.c	2007-02-16 09:53:51.000000000 +0100
+++ bacula.new/src/dird/ua_purge.c	2007-02-20 18:02:46.000000000 +0100
@@ -605,6 +605,18 @@
       }
       pm_strcpy(jcr->VolumeName, mr->VolumeName);
       generate_job_event(jcr, "VolumePurged");
+      if (mr->RecyclePoolId && mr->RecyclePoolId != mr->PoolId) {
+         POOL_DBR oldpr, newpr;
+         memset(&oldpr, 0, sizeof(POOL_DBR));
+         memset(&newpr, 0, sizeof(POOL_DBR));
+         newpr.PoolId = mr->RecyclePoolId;
+         oldpr.PoolId = mr->PoolId;
+         if (db_get_pool_record(jcr, ua->db, &oldpr) && db_get_pool_record(jcr, ua->db, &newpr)) {
+            update_vol_pool(ua, newpr.Name, mr, &oldpr);
+         } else {
+            bsendmsg(ua, "%s", db_strerror(ua->db));
+         }
+      }
       /* Send message to Job report, if it is a *real* job */           
       if (jcr && jcr->JobId > 0) {
          Jmsg1(jcr, M_INFO, 0, _("All records pruned from Volume \"%s\"; marking it \"Purged\"\n"),
